### DISCUSSION TITLE
How to configure SelectionFilter for list-variable models?

### DISCUSSION THREAD

**Discussion by holgerbrandl** [TimefoldAI/timefold-solver#814] [General]
Hi there,

With chained models, one can configure a `SelectionFilter` via  `moveSelectorConfigList.filterClass` for both local search and construction phase. However, with list-variable models, this is not possible:

> java.lang.IllegalArgumentException: Construction Heuristic phase with a list variable does not support moveSelectorConfigList configuration. Remove the moveSelectorConfigList ([ChangeMoveSelectorConfig(null, ValueSelectorConfig(previousTaskOrTool))]) from the config.

What is the intended migration path?

Thanks & Kind regards,
Holger

**Comment by triceo** [comment]
Hello Holger,

the list variable and the chained variable are not functionally equivalent. Not yet anyway. But maybe there are better ways now of accomplishing those same things. 

Would you mind sharing what you were attempting to do?

**Reply by holgerbrandl** [reply]
Thanks Lukas,

we are working on solvers on JSP problems. Typically we need to model only a small subset of machines of being able to perform a specific task (from a sequence of tasks for the different work items being produced). E.g. a drilling task requires a drill and not an oven to give a specific example. Without selection-filters we would face very poor combinatorics (solver performance) because we then would need to rely on constraints (which are by design much slower compared to the filters).

Sticking to the repo examples, envision that in the food packaging quickstart, only some lines are able to package certain food items.

I don't know how to migrate this important bit to list-API (the rest of our first migrated model seem to work fine with it). 

Kind regards 
Holger

**Reply by Christopher-Chianelli** [reply]
I think for list variables, it currently impossible to specify move selectors for CH (and hence the "Construction Heuristic phase with a list variable does not support moveSelectorConfigList" message), but they can be specified in LS.
The solver should spend the majority of the time in LS, where you can specify a `filterClass`.
The "make sure equipment matches task" constraint as described should be very fast, so my approach would be do both constraints and move filters:

- A hard penalty when a task get assigned an equipment that does not match it, so CH will not pick values not assignable to an entity
- A value filter that only select values assignable to a given entity. so LS only checks moves that are feasible

**Reply by holgerbrandl** [reply]
Thanks Christopher.

How can we disable the CH (and initialize the model separately)? 

Is a such filter planned for CH (given the use-cases described above) to bring back the more consistent design option from chain models? (without an additional hard-constraint which only affects CH and is clutter during LS)

I don't agree with the assumption that "make sure equipment matches task" constraint is fast. In particular, in Job-Shop-Scheduling we very very often encounter a complete schedule invalidation after moves because of the interdependencies between operations of each work item. This makes it actually quite costly to evaluate a schedule vs evaluating a selection-filter.

Kind regards
Holger

**Reply by Christopher-Chianelli** [reply]
I am referring to the fact this code will be fast to evaluate:

```java
Constraint makeSureEquipmentMatchesTask(ConstraintFactory cf) {
     return cf.forEach(Task.class)
                     .filter(task -> task.getRequiredEquipment().equals(task.getEquipment()))
                     .penalize(HardSoftScore.ONE_HARD)
                     .asConstraint("Matches equipment");
}
```

The complete schedule invalidation would only occur in CH (which should only be running for a short time), since LS would not pick moves that invalidate the schedule because of an invalid equipment being assigned.

To disable CH/initialize the model separately create a [Custom Solver Phase](https://docs.timefold.ai/timefold-solver/latest/optimization-algorithms/optimization-algorithms#customSolverPhase) and add it to your `solver-config.xml`:

```xml
<solver xmlns="https://timefold.ai/xsd/solver" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="https://timefold.ai/xsd/solver https://timefold.ai/xsd/solver/solver.xsd">
  <!-- ... -->
  <customPhase>
    <customPhaseCommandClass>...MyCustomPhase</customPhaseCommandClass>
  </customPhase>
  <localSearch>
    <unionMoveSelector>
      <listChangeMoveSelector>
        <filterClass>...Filter</filterClass>
          <!-- ... -->
      </listChangeMoveSelector>
      <!-- ... -->
    </unionMoveSelector>
  </localSearch>
</solver>
```

In regards to plans, I agree it is a valid issue; I created https://github.com/TimefoldAI/timefold-solver/issues/823 to track it.

**Reply by holgerbrandl** [reply]
Thanks Christopher for your kind support and guidance.

**Reply by terbrack** [reply]
Hi everyone,
only a short additional note on the above stated blocker:
in a JS scheduling problem with list variables, we model the Variable Listener with the assumption of a valid plan, i.e. task-tool-assignment.
Due to the missing move selector config, invalid task-tool-assignments occur in the construction phase, leading to a score corruption.
Based on that, I am looking forward to a progress on [#823](https://github.com/TimefoldAI/timefold-solver/issues/823).
Best,
Hajo

**Reply by triceo** [reply]
@terbrack If you're seeing move corruptions, I doubt that it is because of missing move filters. Score corruptions come from wrong implementations of variable listeners, custom moves, or constraints. A correct implementation of either of those does not cause score corruptions.

Also, to set the correct expectations here - we are aware of #823, but it is currently not on the roadmap. Timefold Solver is an open source project though; pull requests are welcome.
